### ISSUE TITLE
Regression in padding value and loss calculation

### ISSUE THREAD

**Issue by eegli** [lucidrains/MEGABYTE-pytorch#19]
Hi! I think the bug discussed in #17 was actually not a bug, and that a regression was introduced:

```py
if exists(prev_stage_tokens_repr):
    prev_stage_tokens_repr = F.pad(prev_stage_tokens_repr, (0, 0, 1, 0), value = self.pad_id)
    stage_tokens = stage_tokens + prev_stage_tokens_repr
```

Here, the padding value should always be 0. At this point, `stage_tokens` has already had the start token prepended to it, and if the padding value is anything but 0 - e.g., 2 - then the numerical value 2 is added to the start token of the current stage. I believe the original implementation was correct.

Furthermore, I have a question about the derivation of the loss. After the start token `P_1` is prepended to the flattened sequence (in `logits = torch.cat((start_tokens, logits), dim = -2)`), `preds` may look like this:

`[P_1, 1, 2, 3, 4, 5, ..., n]`

The final slicing, `preds[..., :-1]`, results in `[P_1, 1, 2, 3, 4, 5, ..., n-1]`. Is this correct? Shouldn't `preds` eventually be `[1, 2, 3, 4, 5, ..., n]`?

So that the loss calculation is actually:
```py
loss = F.cross_entropy(
    preds[..., 1:],
    labels,
    ignore_index = self.pad_id
)
```

Maybe I misunderstood something. Thanks for your help!

PS: Is the last part of this check needed?
```py
if exists(next_h_dim) and next_h_dim != dim:
```
because `next_h_dim != dim` will always return `True` as it's an int to tuple comparison.

**Comment by lucidrains** [comment]
@eegli ah thanks Eric for circling back and correcting this

i've also removed that extra conditional; think was trying to save an extra projection if model dimensions were the same between hierarchies, but probably does not hurt to always have it

as for the slicing, if you shifted the predictions to the right, it would be trivial for the network to predict the labels and it would not learn anything

**Comment by eegli** [comment]
Thank you so much for the prompt response and your hard work, really appreciate it!

**Comment by lucidrains** [comment]
no problem! you go make a big discovery
